rust/src/encodings/plain.rs

```diff
@@ -16,37 +16,37 @@
 
 use std::any::TypeId;
 use std::io::{Error, ErrorKind, Read, Result, Seek, SeekFrom};
+use std::marker::PhantomData;
 
-use arrow2::array::{Array, MutableArray, MutablePrimitiveArray, PrimitiveArray};
-use arrow2::array::new_empty_array;
+use arrow2::array::{Array, PrimitiveArray};
 use arrow2::array::Int32Array;
+use arrow2::array::new_empty_array;
 use arrow2::compute::arithmetics::basic::sub_scalar;
 use arrow2::compute::take::take;
+use arrow2::scalar::{PrimitiveScalar, Scalar};
 use arrow2::types::NativeType;
 
 use crate::encodings::Decoder;
 
 /// Plain Decoder
-pub struct PlainDecoder<'a, R: Read + Seek> {
+pub struct PlainDecoder<'a, R: Read + Seek, T: NativeType> {
```

Comment:
Why move return value type from the base class (Decoder) to the concrete class. (PlainDecoder)?

Reply:
In our rust code, `Decode` is a `trait`, not a `class`(or `struct`), and if we implement different decoders like [PLAIN, VAR_BINARY, DICTIONARY](https://github.com/eto-ai/lance/blob/2d9f71c0d073495a4b0462b53ff27f6a13272033/cpp/src/lance/format/schema.cc#L228). `T: NativeType` is only a concept works in PlainDecoder. The common concepts for all decoders should be [Array](https://docs.rs/arrow2/latest/arrow2/array/trait.Array.html) and [Scalar](https://docs.rs/arrow2/latest/arrow2/scalar/trait.Scalar.html), which is already well abstracted in Arrow2.